### PR TITLE
Feature/stream dtos

### PR DESCRIPTION
**Pull request by rma-bryson**

### CODE REVIEW

**Review by MikeNeilson — DISMISSED** [review]

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+@FormattableWith(contentType = Formats.JSONV1, formatter = JsonV1.class, aliases = {Formats.DEFAULT, Formats.JSON})
+@JsonDeserialize(builder = StreamReach.Builder.class)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
+public final class StreamReach implements CwmsDTOBase {
+
+    private final Double upstreamStation;
+    private final Double downstreamStation;
+    private final String comment;
+    private final LocationIdentifier downstreamLocationId;
```
Comment (rma-psmorris) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java]: Re-use your classes that define the locations supporting stationing

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/stream_reach.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  "comment": "This is a comment for the stream reach.",
+  "downstream-location-id": {
+    "office-id": "Office123",
+    "location-id": "Downstream123"
+  },
+  "upstream-location-id": {
+    "office-id": "Office123",
+    "location-id": "Upstream123"
+  },
+  "configuration-id": "Config123"
```
Comment (rma-psmorris) [inline cwms-data-api/src/test/resources/cwms/cda/data/dto/stream_reach.json]: configuration has an office

**Review by rma-psmorris — COMMENTED** [review]
Setup a meeting to go over this DTO with me.

**Comment by rma-psmorris** [comment]
Make sure you define unit of values in  your DTOs

**Review by rma-psmorris — CHANGES_REQUESTED** [review]
See comments

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/stream_reach.json (hunk 2 above)
Comment (rma-bryson) [reply cwms-data-api/src/test/resources/cwms/cda/data/dto/stream_reach.json]: updated configuration ID to LocationIdentifier (will rename locationIdentifier to CwmsIdentifier in a new PR)

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java (hunk 1 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java]: Updated to use StreamNodes for the downstream and upstream locations

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java [hunk 3]
```diff
@@ -35,30 +35,32 @@
...
     private final Double totalDrainageArea;
     private final Double ungagedDrainageArea;
+    private final String areaUnit;
+    private final String stageUnit;
 
     private StreamLocation(Builder builder) {
-        this.streamJunctionIdentifier = builder.streamJunctionIdentifier;
-        this.streamLocationId = builder.streamLocationId;
+        this.streamNode = builder.streamNode;
+        this.streamLocationId = builder.streamId;
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java]: Should keep the variable names consistent between the DTO and the builder.

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    public LocationIdentifier getStreamId() {
+        return streamId;
+    }
+
+    public LocationIdentifier getReachId() {
+        return reachId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java]: Remove equals/hashcode

**Review by adamkorynta — CHANGES_REQUESTED** [review]

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java (hunk 3 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java]: updated streamId to streamLocationId

**Review by adamkorynta — DISMISSED** [review]

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java [hunk 5]
```diff
@@ -48,23 +46,23 @@ public final class StreamReach implements CwmsDTOBase {
...
     @Override
     public void validate() throws FieldException {
-        if (this.reachId == null) {
+        if (this.id == null) {
             throw new FieldException("The 'reachId' field of a StreamReach cannot be null.");
         }
-        reachId.validate();
+        id.validate();
         if (this.streamId == null) {
             throw new FieldException("The 'streamId' field of a StreamReach cannot be null or empty.");
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java:67]: Need to update the error messages

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java [hunk 6]
```diff
@@ -65,10 +65,10 @@ private StreamLocation(Builder builder) {
 
     @Override
     public void validate() throws FieldException {
-        if (this.streamLocationId == null) {
+        if (this.id == null) {
             throw new FieldException("The 'streamId' field of a StreamLocation cannot be null.");
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java]: Error message

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java [hunk 7]
```diff
@@ -37,15 +37,15 @@ private Stream(Builder builder) {
...
         this.comment = builder.comment;
-        this.streamId = builder.streamId;
+        this.id = builder.id;
     }
 
     @Override
     public void validate() throws FieldException {
-        if (this.streamId == null) {
+        if (this.id == null) {
             throw new FieldException("The 'locationIdentifier' field of a Stream cannot be null.");
```
Comment (adamkorynta) [inline cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java]: error message

**Review by adamkorynta — CHANGES_REQUESTED** [review]
missing the JSON updates

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java (hunk 7 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/Stream.java]: updated

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java (hunk 6 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamLocation.java]: updated

File: cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java (hunk 5 above)
Comment (rma-bryson) [reply cwms-data-api/src/main/java/cwms/cda/data/dto/stream/StreamReach.java:67]: updated

**Review by adamkorynta — DISMISSED** [review]

File: cwms-data-api/src/test/java/cwms/cda/data/dto/StreamNodeTest.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+import cwms.cda.formatters.Formats;
+import org.apache.commons.io.IOUtils;
+import org.junit.jupiter.api.Test;
+
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class StreamNodeTest {
```
Comment (rma-psmorris) [inline cwms-data-api/src/test/java/cwms/cda/data/dto/StreamNodeTest.java:39]: add unit

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/stream.json [hunk 9]
```diff
@@ -1,26 +1,30 @@
...
       "name": "UpstreamStream"
     },
     "station": 678.9,
-    "bank": "R"
+    "bank": "R",
+    "station-unit": "ft"
   },
   "length": 10.5,
   "slope": 0.01,
+  "length-unit": "mi",
```
Comment (rma-psmorris) [inline cwms-data-api/src/test/resources/cwms/cda/data/dto/stream.json:23]: You have different units for station and length. That is not supported in the DB API where length and station are in the same units for both retrieval and storage:
 * @param p_length               The length of the stream, in station unit
 ...
 * @param p_station_unit         The unit for stream stationing

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/stream_location.json [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+  "stream-node": {
+    "stream-id": {
+      "office-id": "SPK",
+      "name": "ImOnThisStream"
+    },
+    "bank": "R",
+    "station": 678.9,
+    "station-unit": "ft"
+  },
+  "published-station": 123.45,
```
Comment (rma-psmorris) [inline cwms-data-api/src/test/resources/cwms/cda/data/dto/stream_location.json:15]: unit for station would be inferred by the stream-node?

**Review by rma-psmorris — CHANGES_REQUESTED** [review]
A few more points of feedback

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/stream_location.json (hunk 10 above)
Comment (rma-bryson) [reply cwms-data-api/src/test/resources/cwms/cda/data/dto/stream_location.json:15]: correct. I added some helper methods to stream location to get data off the stream node as well, to get station, station units, and stream id.

File: cwms-data-api/src/test/resources/cwms/cda/data/dto/stream.json (hunk 9 above)
Comment (rma-bryson) [reply cwms-data-api/src/test/resources/cwms/cda/data/dto/stream.json:23]: updated json to reflect DB where length and station are always the same unit

File: cwms-data-api/src/test/java/cwms/cda/data/dto/StreamNodeTest.java (hunk 8 above)
Comment (rma-bryson) [reply cwms-data-api/src/test/java/cwms/cda/data/dto/StreamNodeTest.java:39]: added station unit to tests

**Review by rma-psmorris — APPROVED** [review]
No more feedback

**Review by adamkorynta — APPROVED** [review]
approving after squash

**Pull request merged**